nav2_amcl/src/amcl_node.cpp

```diff
@@ -620,16 +618,24 @@ AmclNode::initialPoseReceived(geometry_msgs::msg::PoseWithCovarianceStamped::Sha
       global_frame_id_.c_str());
     return;
   }
-  // Overriding last published pose to initial pose
-  last_published_pose_ = *msg;
 
-  if (!active_) {
-    init_pose_received_on_inactive = true;
-    RCLCPP_WARN(
-      get_logger(), "Received initial pose request, "
-      "but AMCL is not yet in the active state");
-    return;
+  {
+    std::lock_guard<std::recursive_mutex> cfl(mutex_);
```

Comment:
moving this exposes `global_frame_id` being read without synchronisation above.
Looking at the original upstream code, seems there are a couple of places this is already the case, configure and in the constructor.  The constructor is safe, but I dont see how configure is protected.  activate also assigned `global_frame_id` but this locks the mutex.

Anyhow, if the scope is only to allow `handleInitialPose` to be called without the lock held, consider taking the lock at the start, and just releasing it early

Reply:
err, the not constructor and the lifecycle methods.  Rather the set params callback modifies vars with the lock held, but it seems the lifecycle methods do not